realm/realm-library/src/androidTest/java/io/realm/RealmAsyncQueryTests.java

```diff
@@ -1411,6 +1415,76 @@ public void onChange(RealmResults<AllTypes> results) {
         });
     }
 
+    @Test
+    @UiThreadTest
+    public void executeTransactionAsync_mainThreadQueriesAllowed() {
+        CountDownLatch countDownLatch = new CountDownLatch(1);
+
+        RealmConfiguration configuration = configFactory.createConfigurationBuilder()
+                .allowWritesOnUiThread()
+                .build();
+
+        Realm realm = Realm.getInstance(configuration);
+        realm.executeTransactionAsync(new Realm.Transaction() {
+            @Override
+            public void execute(Realm realm) {
+                realm.insert(new Dog("Snuffles"));
+                countDownLatch.countDown();
+            }
+        });
+
+        TestHelper.awaitOrFail(countDownLatch);
+
+        // Avoid stale data
+        realm.refresh();
+
+        RealmResults<Dog> results = realm.where(Dog.class).equalTo("name", "Snuffles").findAll();
+        assertEquals(1, results.size());
+        assertNotNull(results.first());
+        assertEquals("Snuffles", Objects.requireNonNull(results.first()).getName());
+
+        realm.close();
+    }
+
+    @Test
+    @UiThreadTest
+    public void executeTransactionAsync_throwsWhenRunningOnMainThread() {
```

Comment:
There is some misunderstanding here. Async transactions don't run on the UI thread, they are only started there, but actually run in a background thread, so this test and implementation is wrong. The only thing that should be disallowed is `executeTransaction()` (which runs on the same thread).